File: dixit-companion/src/main/java/fr/erban/dxitcompanion/game/turn/activity/SelectVotesActivity.java
package fr.erban.dxitcompanion.game.turn.activity;

import android.content.Intent;
import android.graphics.Typeface;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.res.ResourcesCompat;

import java.util.ArrayList;
import java.util.List;

import fr.erban.dxitcompanion.R;
import fr.erban.dxitcompanion.game.GameBean;
import fr.erban.dxitcompanion.game.player.PlayerBean;
import fr.erban.dxitcompanion.game.turn.Turn;
import fr.erban.dxitcompanion.game.turn.bean.VoteBean;
import fr.erban.dxitcompanion.game.turn.bean.VotesBean;

public class SelectVotesActivity extends AppCompatActivity {

    private Turn turn;

    private GameBean gameBean;

    private VotesBean votes;

    private PlayerBean voter;

    private List<String> alreadyVoted;

    @Override
    protected void onCreate(Bundle savedInstanceState) {

        super.onCreate(savedInstanceState);
        setContentView(R.layout.select_votes);

        this.turn = (Turn) getIntent().getSerializableExtra("Turn");
        this.gameBean = (GameBean) getIntent().getSerializableExtra("Game");
        this.votes = (VotesBean) getIntent().getSerializableExtra("Votes");

        alreadyVoted = new ArrayList<>();
        alreadyVoted.add(turn.getStoryTeller()
                .getName());
        for (VoteBean vote : votes.getVotes()) {
            alreadyVoted.add(vote.getVoter()
                    .getName());
        }

        if (votes == null) {
            votes = VotesBean.builder()
                    .votes(new ArrayList<>())
                    .build();
        }

        final TextView turnNumber = findViewById(R.id.turnNumber);
        final String turnNumberComplete = getString(R.string.turnNumberPrefix) + gameBean.getCurrentTurn();
        turnNumber.setText(turnNumberComplete);

        reinitActivity();
    }

    private void reinitActivity() {

        List<PlayerBean> players = gameBean.getPlayers();

        for (PlayerBean player : players) {
            boolean hasVoted = false;
            for (String voted : alreadyVoted) {
                if (player.getName()
                        .equals(voted)) {
                    hasVoted = true;
                    break;
                }
            }

            if (!hasVoted) {
                voter = player;
                TextView title = findViewById(R.id.selectVotesTitle);
                String text = "Pour qui a voté " + player.getName() + " ?";
                title.setText(text);
                final Typeface typeface = ResourcesCompat.getFont(this, R.font.write_me_a_song);
                title.setTypeface(typeface);
                title.setTextSize(60);

                if (alreadyVoted.size() == players.size() - 1) {
                    Button nextPlayerBtn = findViewById(R.id.nextPlayerButton);
                    nextPlayerBtn.setVisibility(View.INVISIBLE);
                    Button continueBtn = findViewById(R.id.selectVotesContinueBtn);
                    continueBtn.setVisibility(View.VISIBLE);
                }

                addPlayerNames(gameBean);
                return;
            }
        }
    }

    private void addPlayerNames(GameBean gameBean) {

        final List<PlayerBean> players = gameBean.getPlayers();
        final RadioGroup radioGroup = findViewById(R.id.radioGroupSelectVote);
        int count = radioGroup.getChildCount();
        if (count > 0) {
            for (int i = count - 1; i >= 0; i--) {
                View o = radioGroup.getChildAt(i);
                if (o instanceof RadioButton) {
                    radioGroup.removeViewAt(i);
                }
            }
        }
        radioGroup.clearCheck();

        for (PlayerBean player : players) {
            if (!player.getName()
                    .equals(turn.getStoryTeller()
                            .getName()) && !player.getName()
                    .equals(voter.getName())) {
                RadioButton radioButton = new RadioButton(this);
                radioButton.setText(player.getName());
                radioButton.setTextSize(50);
                final Typeface font = ResourcesCompat.getFont(this, R.font.write_me_a_song);
                radioButton.setTypeface(font);
                radioButton.setTextColor(getResources().getColor(R.color.backgroundTextColor));
                radioGroup.addView(radioButton);
                if (radioGroup.getCheckedRadioButtonId() == -1) {
                    radioButton.setChecked(true);
                }
            }
        }
    }

    public void nextPlayer(View view) {

        alreadyVoted.add(voter.getName());

        ajouterVoteEnCours();

        reinitActivity();
    }

    private void ajouterVoteEnCours() {

        final VoteBean vote = new VoteBean();

        final RadioGroup radioGroup = findViewById(R.id.radioGroupSelectVote);

        final int selectedPlayerId = radioGroup.getCheckedRadioButtonId();
        final View rb = radioGroup.findViewById(selectedPlayerId);
        int idx = radioGroup.indexOfChild(rb);
        final RadioButton radioButton = (RadioButton) radioGroup.getChildAt(idx);
        final String selectedPlayer = radioButton.getText().toString();

        final List<PlayerBean> players = gameBean.getPlayers();

        for (PlayerBean player : players) {
            if (player.getName()
                    .equals(selectedPlayer)) {
                vote.setElected(player);
            }
        }

        vote.setVoter(voter);

        votes.getVotes()
                .add(vote);
    }

    public void continueToEndTurn(View view) {

        ajouterVoteEnCours();

        turn.setVotes(votes.getVotes());

        Intent intent = new Intent(SelectVotesActivity.this, EndTurnActivity.class);
        intent.putExtra("Game", gameBean);
        intent.putExtra("Turn", turn);
        SelectVotesActivity.this.startActivity(intent);
    }
}

File: dixit-companion/src/main/java/fr/erban/dxitcompanion/game/turn/activity/SelectVotesActivity.kt
package fr.erban.dxitcompanion.game.turn.activity

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.RadioButton
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.res.ResourcesCompat
import fr.erban.dxitcompanion.R
import fr.erban.dxitcompanion.common.parcelableExtra
import fr.erban.dxitcompanion.databinding.SelectVotesBinding
import fr.erban.dxitcompanion.game.GameBean
import fr.erban.dxitcompanion.game.player.PlayerBean
import fr.erban.dxitcompanion.game.turn.Turn
import fr.erban.dxitcompanion.game.turn.bean.VoteBean
import fr.erban.dxitcompanion.game.turn.bean.VotesBean

class SelectVotesActivity : AppCompatActivity() {

    private lateinit var binding: SelectVotesBinding
    private lateinit var turn: Turn
    private lateinit var gameBean: GameBean
    private var votes = VotesBean()
    private var voter: PlayerBean? = null
    private val alreadyVoted = mutableListOf<String>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = SelectVotesBinding.inflate(layoutInflater)
        setContentView(binding.root)

        turn = intent.parcelableExtra("Turn")!!
        gameBean = intent.parcelableExtra("Game")!!
        votes = intent.parcelableExtra("Votes") ?: VotesBean()

        alreadyVoted.add(turn.storyTeller!!.name)
        votes.votes.forEach { alreadyVoted.add(it.voter.name) }

        binding.turnNumber.text = getString(R.string.turnNumberPrefix) + gameBean.currentTurn
        reinitActivity()
    }

    private fun reinitActivity() {
        val nextVoter = gameBean.players.firstOrNull { it.name !in alreadyVoted }
        if (nextVoter == null) return
        voter = nextVoter

        binding.selectVotesTitle.apply {
            text = "Pour qui a voté ${nextVoter.name} ?"
            typeface = ResourcesCompat.getFont(this@SelectVotesActivity, R.font.write_me_a_song)
            textSize = 60f
        }

        if (alreadyVoted.size == gameBean.players.size - 1) {
            binding.nextPlayerButton.visibility = View.INVISIBLE
            binding.selectVotesContinueBtn.visibility = View.VISIBLE
        }

        populateRadioGroup()
    }

    private fun populateRadioGroup() {
        val rg = binding.radioGroupSelectVote
        for (i in rg.childCount - 1 downTo 0) {
            if (rg.getChildAt(i) is RadioButton) rg.removeViewAt(i)
        }
        rg.clearCheck()

        gameBean.players
            .filter { it.name != turn.storyTeller?.name && it.name != voter?.name }
            .forEach { player ->
                val rb = RadioButton(this).apply {
                    text = player.name
                    textSize = 50f
                    typeface = ResourcesCompat.getFont(this@SelectVotesActivity, R.font.write_me_a_song)
                    setTextColor(resources.getColor(R.color.backgroundTextColor, theme))
                }
                rg.addView(rb)
                if (rg.checkedRadioButtonId == -1) rb.isChecked = true
            }
    }

    private fun recordCurrentVote() {
        val rg = binding.radioGroupSelectVote
        val rb = rg.findViewById<RadioButton>(rg.checkedRadioButtonId)
        val elected = gameBean.players.first { it.name == rb.text.toString() }
        votes = votes.copy(votes = votes.votes + VoteBean(voter = voter!!, elected = elected))
    }

    fun nextPlayer(view: View) {
        recordCurrentVote()
        alreadyVoted.add(voter!!.name)
        reinitActivity()
    }

    fun continueToEndTurn(view: View) {
        recordCurrentVote()
        turn = turn.copy(votes = votes.votes)
        startActivity(
            Intent(this, EndTurnActivity::class.java)
                .putExtra("Game", gameBean)
                .putExtra("Turn", turn)
        )
    }
}
